Improved code coverage of auth.py

trevvorko opened the pull request:
Added more tests to cover more lines of auth.py. Mainly Authenticate
testing, but also some more missed error handling in CompleteSignUp,
Login, and Signup.

All tests pass, ran `isort` and `black`

Unable to assign reviewers, but I think @aapeliv or @ShmuelTreiger can take a look and let me know of any suggestions or comments.

Part of issue #900

Commit:
Improved code coverage of auth.py

Added more tests to cover more lines of auth.py. Mainly Authenticate
testing, but also some more missed error handling in CompleteSignUp,
Login, and Signup

aapeliv commented on app/backend/src/tests/test_auth.py:
```diff
@@ -309,4 +309,272 @@ def test_signup_invalid_birthdate(db):
+    # Invalid username
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        with pytest.raises(grpc.RpcError) as e:
+            reply = auth_api.Authenticate(auth_pb2.AuthReq(user="Notarealusername", password="password"))
```
Please make username lowercase, otherwise this will fail with the regex.

By the way, it's possible to stick any of three things in the user here: username, email, or user id. Look up `get_user_by_field` or similar

trevvorko replied:
I saw the user id being an option here, but when I try passing that to user I get this error within auth_pb2.AuthReq
`TypeError: 1 has type int, but expected one of: bytes, unicode`

So I was thinking there was some type restriction on what could be passed to AuthReq but I couldn't figure out how to fix it. Do you have any suggestions, or do you even think it's important I test username, email, and user id?

aapeliv replied:
That field has to be passed in as a string, even if you're entering a number, due to the typing of gRPC.

But testing each of the three cases isn't *that* important, up to you if you feel like adding them.

trevvorko replied:
Just as you were responding I found the code that expected it to be a string. You say it's due to gRPC typing, would it be beneficial to add functionality that would accept an int as well? Or would that be just a convenience thing for testing and not actually used in production?

aapeliv replied:
We'd have to add another field, and then have something like 

```proto
message WhateverItsCalled {
  oneof user {
    int64 user_id = 1;
    string user_username = 2;
    string user_email = 3;
  }
}
```

it just moves the burden of regexping from the backend to the frontend. Easier to keep on the backend for now!

I think going forward we might want to get rid of the places where we do this though, user IDs should really never be exposed in the frontend

trevvorko replied:
One final question, is it prefered I add another commit to fix these changes, and then end up merging both of them into develop (if approved), or should I amend the current commit so only one gets merged into develop?

aapeliv replied:
Normally we add commits; it's more convenient for the reviewer to see the new changes, see the progression of CI/CD successes/failures, etc. But you can also amend if you prefer (as long as you only amend in a PR and not on `develop` as rewriting history is bad when multiple people have the branch checked out).

aapeliv commented on app/backend/src/tests/test_auth.py:
```diff
@@ -309,4 +309,272 @@ def test_signup_invalid_birthdate(db):
+
+    # Invalid username
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        reply = auth_api.Login(auth_pb2.LoginReq(user="Notarealusername"))
```
Ditto

aapeliv commented on app/backend/src/tests/test_auth.py:
```diff
@@ -309,4 +309,272 @@ def test_signup_invalid_birthdate(db):
+    assert reply.next_step == auth_pb2.SignupRes.SignupStep.EMAIL_EXISTS
+
+
+def test_successful_authenticate(db):
```
Use the `fast_passwords` fixture in tests that do any password hashing: that will significantly speed the tests up.

aapeliv reviewed (changes requested):
This is looking very good, thanks!

I've added a few small comments, and I'm just running the CI on it.

One small thing, I think it'd probably be better to use non-existent but invalid emails in the tests. We've used `{whatever}@couchers.org.invalid` in a bunch of places. Just to protect in case someone runs the tests with actual SMTP credentials or something...

Commit:
fast_password, invalid email, auth by id to tests

Added to test_auth.py by adding fast password fixture to password
hashing tests. Changed the email being used in tests to an invalid one
to make sure no emails are actually sent in case of an error, and added
auth by id checking to tests.

aapeliv reviewed (approved):
Looks great, thanks!
